k2/csrc/cuda/log.h

```diff
@@ -0,0 +1,185 @@
+// k2/csrc/cuda/log.h
+
+// Copyright (c)  2020  Fangjun Kuang (csukuangfj@gmail.com)
+
+// See ../../../LICENSE for clarification regarding multiple authors
+
+// Glog-like logging functions for k2.
+
+#ifndef K2_CSRC_CUDA_LOG_H_
+#define K2_CSRC_CUDA_LOG_H_
+
+#include <cstdint>
+#include <cstdio>
+
+namespace k2 {
+
+namespace internal {
+
+#if defined(NDEBUG)
+constexpr bool kDisableDebug = true;
+#else
+constexpr bool kDisableDebug = false;
+#endif
+
+enum class LogLevel {
+  kDebug = 0,
+  kInfo = 1,
+  kWarning = 2,
+  kError = 3,
+  kFatal = 4,
+};
+
+// They are used in LOG(xxx), so their names
+// do not follow the google c++ code style
+constexpr LogLevel DEBUG = LogLevel::kDebug;
+constexpr LogLevel INFO = LogLevel::kInfo;
+constexpr LogLevel WARNING = LogLevel::kWarning;
+constexpr LogLevel ERROR = LogLevel::kError;
+constexpr LogLevel FATAL = LogLevel::kFatal;
+
+class Logger {
+ public:
+  __host__ __device__ Logger(const char *filename, const char *func_name,
+                             uint32_t line_num, LogLevel level)
+      : filename_(filename),
+        func_name_(func_name),
+        line_num_(line_num),
+        level_(level) {
+    switch (level) {
+      case DEBUG:
+        printf("[D] ");
+        break;
+      case INFO:
+        printf("[I] ");
+        break;
+      case WARNING:
+        printf("[W] ");
+        break;
+      case ERROR:
+        printf("[E] ");
+        break;
+      case FATAL:
+        printf("[F] ");
+        break;
+    }
+    printf("%s:%s:%u ", filename, func_name, line_num);
+#if defined(__CUDA_ARCH__)
+    printf("block:[%u,%u,%u], thread: [%u,%u,%u] ", blockIdx.x, blockIdx.y,
+           blockIdx.z, threadIdx.x, threadIdx.y, threadIdx.z);
+#endif
+  }
+
+  __host__ __device__ ~Logger() {
+    printf("\n");
+    if (level_ == FATAL) {
+#if defined(__CUDA_ARCH__)
+      // this is usually caused by one of the CHECK macros and the detailed
+      // error messages should have already been printed by the macro, so we
+      // use an arbitrary string here.
+      __assert_fail("Some bad things happened", filename_, line_num_,
```

Comment:
just to make sure, Is `__assert_fail` supposed to be called by users of cuda as it's prefix with `__`? I know there's a function `assert` in cuda that can be called in kernel.

Reply:
`assert` is guarded by `NDEBUG` macro. I have tried it before but it expands to empty in the release build.

Not sure if the check should be in effect only in the release build or not. I can switch to `assert` if we only
need this check in the release build.

Reply:
From `/usr/local/cuda/include/crt/common_functions.h`

```cpp
#ifdef NDEBUG
#define assert(e) (static_cast<void>(0))
#else /* !NDEBUG */
#define __ASSERT_STR_HELPER(x) #x
#define assert(e) ((e) ? static_cast<void>(0)\
                       : __assertfail(__ASSERT_STR_HELPER(e), __FILE__,\
                                      __LINE__, __PRETTY_FUNCTION__,\
                                      sizeof(char)))
#endif /* NDEBUG */
```

Reply:
I just worried about the Nvidia may change the name of `__asserfail` later if it's not supposed to be caller by normal users. Or Can we add a TODO or Caution here, and we can check this later in the future if we find better solution

Reply:
I guess `__asserfail`  is a standard name, you can also find it in `/usr/include/assert.h`.

From <https://refspecs.linuxbase.org/LSB_3.0.0/LSB-Core-generic/LSB-Core-generic/baselib---assert-fail-1.html>

> The __assert_fail() function is not in the source standard; it is only in the binary standard

I am not sure which **binary standard** it refers to. But I am confident that the function
signature will be stable in the future.

Reply:
OK. It may refer to ABI here?

Reply:
Maybe...

I just found `__assertfail` from CUDA's own documentation: <https://docs.nvidia.com/cuda/ptx-writers-guide-to-interoperability/index.html>

> In order to support assert, the PTX function call __assertfail is used whenever the assert expression produces a false value. The prototype for the __assertfail system call is: ....
>
> The __assertfail system call is emitted as part of the assert macro defined in "assert.h".
>
> System calls are calls into the driver operating system code. In PTX they look like regular calls, but the function definition is not given. A prototype must be provided in the PTX file, but the implementation of the function is provided by the driver.

So I think as long as `__assertfail` is not changed in the host side, NVIDIA will keep the current signature.

Reply:
Yes, I guess so, thanks, let's keep it.

Reply:
agreed.  We may still want LOG(FATAL) things that die in release mode.

Reply:
Hi, just found this thread, I think for an hour about it.

The needing here actually is we want to a valid assert when we want, like by call `K2_PARANOID_ASSERT`.

Thus, the barrier is `NDEBUG`. I could undef it and define it like below. Do you guys think it's  ok?
```cpp
#ifdef K2_PARANOID
#pragma push_macro("NDEBUG")
#undef NDEBUG
#include <cassert>
#pragma pop_macro("NDEBUG")
#endif

#ifdef K2_PARANOID
#define K2_PARANOID_ASSERT \
		... 
		assert \
```

Reply:
A test code said ok:
```cpp
#define K2_PARANOID

#ifdef K2_PARANOID
#pragma push_macro("NDEBUG")
#undef NDEBUG
#include <cassert>
#pragma pop_macro("NDEBUG")
#endif
#include <cstdio>

#ifdef K2_PARANOID
#define K2_PARANOID_ASSERT(exp, format, ...)                                \
  do {                                                                      \
    if (exp)                                                                \
      (void)0;                                                              \
    else {                                                                  \
      assert(exp);                                                          \
    }                                                                       \
  } while (0)
#else
#define K2_PARANOID_ASSERT(exp, format, ...) ((void)0)
#endif


__global__ void testAssert(void) {
  int is_one = 1;
  int should_be_one = 0;

#ifdef NDEBUG
  printf("NDEBUG is still here.\n");
#endif

  // This will have no effect
  // This will halt kernel execution
  K2_PARANOID_ASSERT(should_be_one, "");

  assert(should_be_one);
}

int main() {
  testAssert<<<1, 1>>>();
  cudaDeviceSynchronize();
  auto e = cudaGetLastError();
  if (e != cudaSuccess) {
    printf("cuda error: [%s] %s.\n",
           cudaGetErrorName(e), cudaGetErrorString(e));
    exit(e);
  }
  return 0;
}
```

---

![image](https://user-images.githubusercontent.com/705582/92321941-c8866900-f060-11ea-87fc-6708a345e7cc.png)